Google Authentication for PA (CU-3denbe6)

micahdbak opened the pull request:
# Google Authentication for PA

This PR is the last of a series of features for `brave-alert-lib`, `BraveSensor`, and `BraveButtons` that implements Google authentication as the primary means of authorization for PA and PA's API routes.

## This PR...

### Added

- Google login screen and page (CU-3denbe6).

### Changed

- Renamer to authenticate with the PA API using Google (CU-3denbe6).
- Button Registration to authenticate with the PA API using Google (CU-3denbe6).
- Twilio Number Purchasing to authenticate with the PA API using Google (CU-3denbe6).

The primary goal of using Google to authenticate in PA is to secure PA API routes that don't depend on Particle or ClickUp's login credentials. The list of changed PA API routes can be seen below.

## PA API Routes

### New

#### BraveSensor

- /pa/get-google-tokens
- /pa/get-google-payload

### Changed

The following PA API routes now require that `googleIdToken` is posted in the body of the request instead of `clickUpToken`.

#### BraveSensor

- /pa/create-sensor-location
  - Used by Renamer in PA
- /pa/get-sensor-clients
  - Used by Renamer in PA
- /pa/sensor-twilio-number
  - Used by Twilio Number Purchasing in PA

#### BraveButtons

- /pa/aws-device-registration
  - Used by Button Registration in PA
- /pa/buttons-twilio-number
  - Used by Twilio Number Purchasing in PA

## Test Plan:
- [x] Existing functionality of PA works as expected
  - [x] Device Manager
  - [x] Activator
  - [x] Device Lookup
  - [x] Door Sensor Pairing
  - [x] Renamer
  - [x] Button Registration
  - [x] Twilio Number Purchasing
- [x] Changed views authorize with Google, not ClickUp
  - [x] Renamer
  - [x] Button Registration
  - [x] Twilio Number Purchasing

micahdbak committed:
Added Google Log In for PA

micahdbak committed:
Added Google profile page to Navigation

micahdbak committed:
Stored Google access token in cookies

micahdbak committed:
Merge branch 'main' into mdb-google-oauth-login

trisapeace reviewed (commented):
This all looks good to me @micahdbak . But as we talked about earlier today, the API calls to the Sensors and Buttons servers also need to be protected, not just the UI. I imagine it will be something like sending the googleAccessToken in the API calls to `/pa/sensor-twilio-number`, `/pa/buttons-twilio-number`, `/pa/get-sensor-clients`, `/pa/create-sensor-location`, and `/pa/aws-device-registration`. 

Then on the server side, you'll need to verify that the sent googleAccessToken is valid. Maybe using something like this: https://developers.google.com/identity/gsi/web/guides/verify-google-id-token. 

Right now we're sending the clickUpToken to each of these API calls and verifying it using `clickUpHelpers.clickUpChecker`. You can probably take this out and use the Google version instead. Although it does look as though the `clickUpHelpers.clickUpChecker` (in BraveAlertLib) is making sure that the ClickUpToken has the right name and team ID before it authorizes the request. Maybe the security issue isn't as bad as you originally thought?

micahdbak commented:
> This all looks good to me @micahdbak . But as we talked about earlier today, the API calls to the Sensors and Buttons servers also need to be protected, not just the UI. I imagine it will be something like sending the googleAccessToken in the API calls to `/pa/sensor-twilio-number`, `/pa/buttons-twilio-number`, `/pa/get-sensor-clients`, `/pa/create-sensor-location`, and `/pa/aws-device-registration`.

I'll start working on this then!

> Then on the server side, you'll need to verify that the sent googleAccessToken is valid. Maybe using something like this: https://developers.google.com/identity/gsi/web/guides/verify-google-id-token.
> 
> Right now we're sending the clickUpToken to each of these API calls and verifying it using `clickUpHelpers.clickUpChecker`. You can probably take this out and use the Google version instead. Although it does look as though the `clickUpHelpers.clickUpChecker` (in BraveAlertLib) is making sure that the ClickUpToken has the right name and team ID before it authorizes the request. Maybe the security issue isn't as bad as you originally thought?

Also, I just looked into this, and you're definitely right.
I just created a ClickUp account from scratch and wasn't able to purchase a Twilio number (what I suspected was the prime security risk), so the issue is not as bad as I thought at all 🤣.

That said, I'll keep working on Google being the single form of authentication for PA for the purpose of simplicity!

trisapeace commented:
Cool. Thanks. Yes, I think that Google will be better than ClickUp because there is the known issue that ClickUp always returns the same token every time. So if anyone did manage to get our token, then we'd be compromised. Whereas Google tokens should expire after a certain period of time. It might even be possible to specify the timeout when we request a token and then you can pick a length of time that feels appropriate for this application (maybe 8 hours to give us one day's worth of time without needing to login again).

micahdbak commented:
Yeah, specifying a timeout would be great. I'll look into that! 😊

micahdbak committed:
Reworked Google auth to occur in PA API

micahdbak committed:
Migrated ClickUp authentication for sensor PA API calls

trisapeace reviewed (commented):
This is looking great! No particular comments, but I don't want to approve until you're done and I've taken a look your test plan.

micahdbak committed:
Updated Google Login Screen & Usage of Google ID Token

trisapeace commented:
Task linked: [CU-3denbe6 Implement Google authentication for PA](https://app.clickup.com/t/3denbe6)

micahdbak committed:
Merge branch 'main' into mdb-google-oauth-login

micahdbak committed:
Added display message when waiting for server response

micahdbak committed:
Imported axios properly for axios v1.6.X

See https://www.npmjs.com/package/axios#example for importing axios.

- Updated 'npm run lint' to lint only the src/ directory.
  - This is to prevent waiting n extra minutes when you have run `npm
    run deploy`, or similar, and eslint attempts to lint very large
    binary files.

micahdbak committed:
Cleaned up GoogleLoginScreen.jsx

micahdbak committed:
Fixed bug with Clickup login attempting multiple times

micahdbak committed:
Updated CHANGELOG.md

tmackinnon commented on src/upper-level-components/GoogleLoginScreen.jsx:
```diff
@@ -0,0 +1,104 @@
+  })
+
+  // attempt to log in if a Google ID token exists in cookies
+  if (cookies.googleIdToken !== undefined) {
+    handleGoogleIdToken(cookies.googleIdToken)
```
Do you mind explaining why you didn't use the `await` keyword in front of `handleGoogleIdToken`?

micahdbak replied:
I chose to let the handleGoogleIdToken function run asynchronously as I'm not sure if a React component can return a Promise... what do you think?

At present, given a cookie already exists, the handleGoogleIdToken function starts running, and then an empty div element is returned. When handleGoogleIdToken completes, it will update one of the component's states, or the googleIdToken cookie, after which the entire component will refresh.

tmackinnon replied:
Ohhh I see what you mean, I'm thinking it might be better to use a useEffect hook in this case. My understanding is that in React useEffect is used to handle "side effects" like fetching data. So I'm thinking something like this
```
useEffect(() => {
    const loginWithToken = async () => {
      if (cookies.googleIdToken !== undefined) {
        await handleGoogleIdToken(cookies.googleIdToken);
      }
    };

    loginWithToken();
  }, [cookies.googleIdToken]); // This dependency will ensure this re-runs if the googleIdToken changes 
 ```
  
So the useEffect will run when the component renders, and then run again if the googleIdToken changes. What do you think?

micahdbak replied:
Hmm, upon doing a little research, I cannot use `useEffect` with a cookie being in the dependency list, it must be a state created with `useState`. I can add a `googleIdToken` state, but I feel that is unnecessary overhead when a cookie already exists... I'll try and wrap the existing code in an effect, but I'm a little worried that the login screen will appear for a split second when googleIdToken exists in the cookies. This is why I am returning an empty div in the existing check.

micahdbak replied:
OK, I've implemented a solution that I'm happy with, and it uses effects. Let me know your thoughts!

tmackinnon replied:
Ahh shoot, I was thinking the useCookies worked similarly to useState but I see what you mean, and I agree it doesn't make sense to add `googleIDToken` into state on top of it existing as a cookie. 

I like the solution you came up with! Glad you were able to wrap it in a useEffect and cover your bases with return the empty div.

tmackinnon commented on src/views/ClickupLogin.jsx:
```diff
@@ -9,6 +9,7 @@ function ClickupLogin(props) {
   // Fetch OAuth code from redirect URL
   const urlParams = new URLSearchParams(window.location.search)
   const clickupCode = urlParams.get('code')
+  const [performingTokenLogin, setPerformingTokenLogin] = useState(false)
```
For my own understanding, do you mind explaining why performingTokenLogin was added here and used below? Thanks!

micahdbak replied:
I added the performingTokenLogin state here to resolve a weird bug Johnny encountered. Essentially, upon attempting a ClickUp login, this component would attempt to log into ClickUp several times. (This didn't occur in the previous version of PA.) This might be because my additions / modifications to a parent component cause the Frame component to refresh multiple times, perhaps because of cookie updates, thus rerunning the ClickUp login code before any of the previous runs have completed. The code here started the login process asynchronously, but didn't check if it was already performing the login. My modification ensures that it only attempts the login once, and waits until that single attempt is finished before reattempting login.

tmackinnon replied:
Ohh interesting! I wonder if you added an empty array as a dependency to the useEffect, if that would help? I don't know if this is the exact same case, but the react the [documentation](https://react.dev/reference/react/useEffect#reference) says if you omit the dependency array, the useEffect will re-run after every re-render of the component. And there doesn't seem to be a dependency array here. But this wouldn't really explain why it was working with no issues before haha.

micahdbak replied:
Cool!!! I didn't realize the empty array did this. I'll implement this instead of my performingTokenLogin logic :D

tmackinnon commented on src/utilities/AWSFunctions.js:
```diff
@@ -18,7 +18,7 @@ const BUTTONS_STAGING_URL = process.env.REACT_APP_BUTTONS_STAGING_URL
  * @return {Promise<string|*>} 'success' if successful, server error message if not.
```
I can't seem to highlight the exact line in my review but just a reminder to update line 17 `* @param {string} clickupToken Global clickup token for authentication` to reflect the use of a googleIdToken instead of a clickupToken.

micahdbak replied:
Yay! Good catch, thank you!

tmackinnon commented on src/utilities/TwilioFunctions.js:
```diff
@@ -24,7 +24,7 @@ const BRAVE_API_KEY_PROD = process.env.REACT_APP_BRAVE_API_KEY_PROD
  * @param {string} clickupToken      clickup auth token
```
Another reminder to update this to the `googleIdToken` instead of `clickupToken`. :)

micahdbak replied:
Will do!

tmackinnon commented on src/utilities/TwilioFunctions.js:
```diff
@@ -61,7 +61,7 @@ async function purchaseTwilioNumberByAreaCode(url, areaCode, locationID, environ
  * @param {string} clickupToken  clickup auth token
```
Another reminder to update this to the googleIdToken instead of clickupToken. :)

micahdbak replied:
Will do!

tmackinnon commented on src/views/ButtonRegistration.jsx:
```diff
@@ -1,25 +1,27 @@
 
 export default function ButtonRegistration(props) {
   // eslint-disable-next-line no-unused-vars
-  const { clickupToken, environment } = props
+  const { environment } = props
 
   const [deviceEUI, setDeviceEUI] = useState('')
   const [deviceName, setDeviceName] = useState('')
   const [formLock, setFormLock] = useState(false)
   const [registrationStatus, setRegistrationStatus] = useState('idle')
+  const [cookies] = useCookies(['googleIdToken'])
```
Cool! I like that we're able to use cookies instead of passing the `googleIdToken` via props.

tmackinnon reviewed (commented):
Wow, it's so great to see this all come together, great job Micah! After playing around with this in the Dev environment, it seems to be working as expected which is awesome. I'm glad you've tested the functionality with Johnny as well.

I just had a couple of reminders and had some clarifying questions for my own understanding but otherwise I think it looks good!

micahdbak committed:
Replaced comments for cickupToken with googleIdToken

micahdbak committed:
Reworked on-load logic for GoogleLoginScreen and ClickupLogin

micahdbak committed:
Oops

micahdbak committed:
Fixed linting errors

tmackinnon reviewed (approved):
Thanks for addressing my comments from earlier, it looks good on my end and is working in the Dev environment as expected! I'm happy to approve this. Great work!!

trisapeace commented on src/upper-level-components/Pages.js:
```diff
@@ -19,6 +19,15 @@ const Pages = {
+    displayName: 'Google',
+    paths: ['/google'],
+    authorizations: {
+      clickup: false,
```
Now that you've removed the need for a ClickUp token for Button Registration, Twilio Number Purchasing, and Renamer, do you think that those pages can have `clickup: false` now?

micahdbak replied:
Good point! I'll do this!

micahdbak replied:
Renamer still needs clickup, but Button Registration and Twilio Number Purchasing don't. I've set clickup to false for those pages.

trisapeace commented on package.json:
```diff
@@ -35,7 +37,7 @@
     "test": "react-scripts test",
     "eject": "react-scripts eject",
-    "lint": "eslint ."
+    "lint": "eslint src"
```
Was there a specific file that you were trying to avoid the linter? If so, I think you can add exclusions to the `.eslintrc.json` file.

micahdbak replied:
Oh, I didn't know I could do that! I added this because it started to lint the `build/` folder when I ran npm run build, which gave me a kajillion warnings and errors... I decided to change this to 'src' because I didn't think there were any files in the root directory that need to be linted, but I think excluding the `build/` folder is a better solution. Thanks!

trisapeace commented on src/pdf/BraveLogoHorizontalSlate.png:
What do you think about moving this file into `src/graphics/`? I think that `src/pdf/` is used for the PDF labels generated in the Renamer page, whereas you're using this for the Google Login, right?

trisapeace replied:
I like the look of the branded login screen, by the way. Nice, clear, and clean.

micahdbak replied:
Ah, yeah, I wondered why there was different folders. Gotcha!

trisapeace commented on src/upper-level-components/Navigation.jsx:
```diff
@@ -34,6 +35,10 @@ function Navigation(props) {
   const loginButtonConfig = {}
 
+  loginButtonConfig[Pages.google.displayName] = {
+    loginState: 'true',
```
Just curious if this needs to be a string as opposed to the boolean `true`?

micahdbak replied:
Good point... I'll change this.

micahdbak replied:
Ah, yes, I remember why this is a string. The loginState value is used in the StatusBadge component under src/components/general/StatusBadge.jsx, where it checks against the following strings to determine state: 'idle', 'waiting', 'true', 'notSelected', 'fail', and else.

trisapeace replied:
Thanks!

trisapeace commented on src/utilities/ClickupFunctions.js:
```diff
@@ -1,8 +1,8 @@
+import axios from 'axios'
```
Sweet.

trisapeace commented on src/views/ClickupLogin.jsx:
```diff
@@ -23,7 +23,7 @@ function ClickupLogin(props) {
       tokenLogin()
     }
-  })
+  }, [])
```
Good catch.

trisapeace reviewed (approved):
Nice work! I've added some comments for things that you can change if you'd like. 

Thanks for testing all the functionality with Johnny!

micahdbak committed:
Updates as per Theresa's review

- Reset `npm run lint` to `eslint .`; updated `.eslintrc.json` instead.
- Moved non-PDF graphics from `src/pdf` to `src/graphics`.
- Removed ClickUp dependency from pages that no longer need ClickUp.
